### PR TITLE
Bucket sort

### PR DESCRIPTION
**Pull request by santushtisharma10**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

I have added python file of Bucket Sorting Algorithm and updated the readme file also.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)

### CODE REVIEW

**Comment by santushtisharma10** [comment]
@SKAUL05 Please review my PR

File: Python/sort/Bucket_Sort.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            index2 -= 1
+            
+        array[index2+1] = value
+      
+    return array   
+
+# Function to perform Bucket Sorting
+def BucketSort(unsortedArray, size):
+    
+    Bucket = []
```
Comment (SKAUL05) [inline Python/sort/Bucket_Sort.py]: Follow single convention for variable naming. Will suggest to follow PEP8 Guidelines
https://www.python.org/dev/peps/pep-0008/

File: Python/sort/Bucket_Sort.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        
+    return SortedArray
+      
+# Input the size of Unsorted List/Array
+size = int(input())
+
+unsortedArray = []
+
+# Input the elements of List/Array
+for index in range(size):
```
Comment (SKAUL05) [inline Python/sort/Bucket_Sort.py]: Include a main function and then call the respective functions from there.

**Review by SKAUL05 — CHANGES_REQUESTED** [review]

**Comment by santushtisharma10** [comment]
@SKAUL05 I have made changes as you said, Please review it

**Review by SKAUL05 — APPROVED** [review]

**Comment by SKAUL05** [comment]
@santushtisharma10 Copied from [here](https://www.geeksforgeeks.org/bucket-sort-2/)

- Exactly same, in some places only variable name is changed.
- Don't just copy & paste.
- Do it by your own
Note: It's very important to do legit work on your own. In case of repeated violations, you will be blocked to contribute in [Neo Algo](https://github.com/TesseractCoding/NeoAlgo)

**Comment by santushtisharma10** [comment]
@SKAUL05 the code of geeks for geeks works only for numbers between 0 and 1 and my code works for large as well as small numbers

**Comment by santushtisharma10** [comment]
I used these as referece https://www.youtube.com/watch?v=geVyIsFpxUs&t=355s and https://en.wikipedia.org/wiki/Bucket_sort

**Comment by santushtisharma10** [comment]
Moreover in geeks for geeks code they always used 10 buckets and since i took the help of wikipedia to understand the algorithm i created the buckets of the same size of unsorted array

**Comment by santushtisharma10** [comment]
@SKAUL05 please look into this matter

File: Python/sort/Bucket_Sort.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    return array   
+
+# Function to perform Bucket Sorting
+def bucket_sort(unsortedArray, size):
+    
+    bucket = []  
+    MAX = max(unsortedArray)
+    
+    # Bucket should be of same size as that of Unsorted Array
+    for index in range(size):
```
Comment (SKAUL05) [inline Python/sort/Bucket_Sort.py]: This could have been easily achieved by 
`bucket = [[]]*size`

File: Python/sort/Bucket_Sort.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+# Bucket Sort Algorithm
+
+# Function to perform Insertion Sort
+def insertion_sort(array, size):
```
Comment (SKAUL05) [inline Python/sort/Bucket_Sort.py:4]: Placing of statements in insertion sort function is exact replica of GFG code only with minor changes in variable names.

**Review by SKAUL05 — COMMENTED** [review]
There are these elements which helps in establishing strong relation between the two scripts which can't be neglected as coincidence.
Since the program is modified in a better way, I think it can be taken as a contribution but will require approval from PA as well.
Do keep in mind while raising PR, if you think you have taken references **(not copied)** from certain sites. Do mention it in PR Description as it will help in avoiding unnecessary element of suspicious activity.
I'm going ahead and opening it for thorough review from other reviewers.

**Comment by santushtisharma10** [comment]
@SKAUL05 thank you i will keep it in mind

**Comment by SKAUL05** [comment]
@HarshCasper @atarax665 @Kajol-Kumari 
Review Required on this PR for plagiarism check and modifications.
Thanks.

**Review by Abhijit2505 — CHANGES_REQUESTED** [review]
First thing first, you have obviously taken reference from the outer source.

I would like you to do 2 things,

1 - make the code shorter by using the reviews given above by @SKAUL05.
2 - Add some modularity to your code, i.e. please add two **class**, inside which you will have your functions. Next you can create custom objects and write the code again. This will solve my query on plagiarism.

**Comment by santushtisharma10** [comment]
@Abhijit2505 Sure

File: Python/sort/Bucket_Sort.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    for index in range(size):
+        
+        if len(bucket[index]):
+            sortedArray.extend(insertion_sort(bucket[index], len(bucket[index])))
+        
+    return sortedArray
+
+def main():
+    # Input the size of Unsorted List/Array
+    size = int(input())
```
Comment (Kajol-Kumari) [inline Python/sort/Bucket_Sort.py]: please show proper message to the user while taking the input.

File: Python/sort/Bucket_Sort.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    unsortedArray = []
+
+    # Input the elements of List/Array
+    for index in range(size):
+    
+        element = int(input())
```
Comment (Kajol-Kumari) [inline Python/sort/Bucket_Sort.py]: same here. please show proper message to the user.

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
Hey @santushtisharma10 I ran your code and the expected output isn't equal to what i am getting here. Please correct this. Also please add one sample input-output below the code section in commented form.
```
4                                                                                                                               
1                                                                                                                               
5                                                                                                                               
6                                                                                                                               
2                                                                                                                               
Unsorted Array : [1, 5, 6, 2]                                                                                                   
Sorted Array :  [1, 2, 5, 6, 1, 2, 5, 6, 1, 2, 5, 6, 1, 2, 5, 6] 
```

**Pull request closed without merging**